src/hotspot/share/gc/g1/heapRegionRemSet.hpp

```diff
@@ -339,74 +104,48 @@ class HeapRegionRemSet : public CHeapObj<mtGC> {
   bool is_updating() { return _state == Updating; }
   bool is_complete() { return _state == Complete; }
 
-  void set_state_empty() {
-    guarantee(SafepointSynchronize::is_at_safepoint() || !is_tracked(), "Should only set to Untracked during safepoint but is %s.", get_state_str());
-    if (_state == Untracked) {
-      return;
-    }
-    clear_fcc();
-    _state = Untracked;
-  }
-
-  void set_state_updating() {
-    guarantee(SafepointSynchronize::is_at_safepoint() && !is_tracked(), "Should only set to Updating from Untracked during safepoint but is %s", get_state_str());
-    clear_fcc();
-    _state = Updating;
-  }
-
-  void set_state_complete() {
-    clear_fcc();
-    _state = Complete;
-  }
-
-  void add_reference(OopOrNarrowOopStar from, uint tid) {
-    RemSetState state = _state;
-    if (state == Untracked) {
-      return;
-    }
-
-    uint cur_idx = _hr->hrm_index();
-    uintptr_t from_card = uintptr_t(from) >> CardTable::card_shift;
+  inline void set_state_empty();
+  inline void set_state_updating();
+  inline void set_state_complete();
 
-    if (G1FromCardCache::contains_or_replace(tid, cur_idx, from_card)) {
-      assert(contains_reference(from), "We just found " PTR_FORMAT " in the FromCardCache", p2i(from));
-      return;
-    }
-
-    _other_regions.add_reference(from, tid);
-  }
+  inline void add_reference(OopOrNarrowOopStar from, uint tid);
 
   // The region is being reclaimed; clear its remset, and any mention of
   // entries for this region in other remsets.
   void clear(bool only_cardset = false);
   void clear_locked(bool only_cardset = false);
 
+  G1CardSetMemoryStats card_set_memory_stats() const { return _card_set_mm.memory_stats(); }
+
   // The actual # of bytes this hr_remset takes up.
   // Note also includes the strong code root set.
   size_t mem_size() {
-    MutexLocker x(&_m, Mutex::_no_safepoint_check_flag);
-    return _other_regions.mem_size()
+    return _card_set.mem_size()
       // This correction is necessary because the above includes the second
       // part.
-      + (sizeof(HeapRegionRemSet) - sizeof(OtherRegionsTable))
+      + (sizeof(HeapRegionRemSet) - sizeof(G1CardSet))
```

Comment:
Pre-existing, but what do you think about moving/including the comment for this part of the calculation in the function comment.

Reply:
I think the comment should stay where it is - it is confusing otherwise, and establishing the context for this seems hard. I can remove the comment if you want though.

Reply:
Removing sounds good to me :) Or fit it on one line like this:
```
      + (sizeof(HeapRegionRemSet) - sizeof(G1CardSet)) // Avoid double counting G1CardSet
```